### PR TITLE
TG-27: create mothods for add and edit admins

### PR DESCRIPTION
**Pull request by EmanuelMuises95**
Create method for add admin
Create method for update admin

### CODE REVIEW

File: src/Components/Admins/Form/index.js [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+            />
+          </div>
+          <input
+            type="button"
+            value="Submit"
+            onClick={adminId ? () => editAdmin() : () => createAdmin()}
+          />
+        </form>
+      </div>
+      {modalDisplay ? (
```
Comment (lmalvasia) [inline src/Components/Admins/Form/index.js:147]:
```suggestion
      {showModal && (
```

File: src/Components/Admins/Form/index.js [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+        <Modal
+          title={modalTitle}
+          contentMessage={contentMessage}
+          setModalDisplay={setModalDisplay}
+        />
+      ) : null}
```
Comment (lmalvasia) [inline src/Components/Admins/Form/index.js:153]:
```suggestion
      )}
```

File: src/Components/Admins/Modal/modal.js [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+      <div id="id-modal" className={modalstyles.modal}>
+        <header className={modalstyles.header}>
+          <h3 className={modalstyles.header__title}>{title}</h3>
+          <button className={modalstyles.header__button} onClick={() => setModalDisplay(false)}>
+            <i className="fa-solid fa-square-xmark fa-2xl"></i>
+          </button>
+        </header>
+        <div className={modalstyles.content}>
+          {content ?? null}
+          {contentMessage ? <p className={modalstyles.content__message}>{contentMessage}</p> : null}
```
Comment (lmalvasia) [inline src/Components/Admins/Modal/modal.js:21]:
```suggestion
          {contentMessage && <p className={modalstyles.content__message}>{contentMessage}</p>}
```

File: src/Components/Admins/Modal/modal.js [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+      ></div>
+      <div id="id-modal" className={modalstyles.modal}>
+        <header className={modalstyles.header}>
+          <h3 className={modalstyles.header__title}>{title}</h3>
+          <button className={modalstyles.header__button} onClick={() => setModalDisplay(false)}>
+            <i className="fa-solid fa-square-xmark fa-2xl"></i>
+          </button>
+        </header>
+        <div className={modalstyles.content}>
+          {content ?? null}
```
Comment (lmalvasia) [inline src/Components/Admins/Modal/modal.js:20]: You are not using the content prop in the form, so we could remove it.

File: src/Components/Admins/Modal/modal.js [hunk 5]
```diff
@@ -0,0 +1,34 @@
+import React from 'react';
+import modalstyles from './modal.module.css';
+
+const Modal = ({ content, contentMessage, title, setModalDisplay }) => {
+  return (
+    <>
+      <div
```
Comment (lmalvasia) [inline src/Components/Admins/Modal/modal.js:7]: This div is empty, we could remove it.

File: src/Components/Admins/Modal/modal.js [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+          <button className={modalstyles.header__button} onClick={() => setModalDisplay(false)}>
+            <i className="fa-solid fa-square-xmark fa-2xl"></i>
+          </button>
+        </header>
+        <div className={modalstyles.content}>
+          {content ?? null}
+          {contentMessage ? <p className={modalstyles.content__message}>{contentMessage}</p> : null}
+          <button
+            className={`${modalstyles.options__button} ${modalstyles.options__close}`}
+            onClick={() => setModalDisplay(false)}
```
Comment (lmalvasia) [inline src/Components/Admins/Modal/modal.js:24]: We should redirect the user to the "list" page when the request is correct. You should check the title of the modal to do this.
```suggestion
            onClick={() => {
              setModalDisplay(false);
              if (title !== 'Error') {
                 window.location.href = '/admins';
              }
            }}
```

File: src/Components/Admins/Form/index.js [hunk 7]
```diff
@@ -0,0 +1,158 @@
+import styles from './form.module.css';
+import { useState, useEffect } from 'react';
+import Modal from '../Modal/modal';
+
+const Form = () => {
+  const paramsURL = new URLSearchParams(window.location.search);
+  const adminId = paramsURL.get('id');
+  const [adminCreate, setAdminCreated] = useState({
```
[the remark this answers is not in the document]
Comment (EmanuelMuises95) [reply src/Components/Admins/Form/index.js]: Done

File: src/Components/Admins/Form/index.js [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+    name: '',
+    lastName: '',
+    email: '',
+    password: ''
+  });
+  const [modalDisplay, setModalDisplay] = useState('');
```
[the remark this answers is not in the document]
Comment (EmanuelMuises95) [reply src/Components/Admins/Form/index.js]: Done

File: src/Components/Admins/Form/index.js [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+  const [contentMessage, setContentMessage] = useState('');
+  const [modalTitle, setModalTitle] = useState('');
+
+  useEffect(async () => {
+    if (adminId) {
+      try {
+        const res = await fetch(`${process.env.REACT_APP_API_URL}/admins/${adminId}`);
+        const data = await res.json();
+        setAdminCreated({
```
[the remark this answers is not in the document]
Comment (EmanuelMuises95) [reply src/Components/Admins/Form/index.js]: Done

File: src/Components/Admins/Form/index.js [hunk 10]
```diff
@@ -0,0 +1,158 @@
...
+        body: JSON.stringify(adminCreate)
+      });
+      const data = await response.json();
+      setContentMessage(data.message);
+      if (response.ok) {
+        setModalTitle('Success');
+      } else {
+        setModalTitle('Error');
+      }
+      setModalDisplay(true);
```
[the remark this answers is not in the document]
Comment (EmanuelMuises95) [reply src/Components/Admins/Form/index.js]: Done

File: src/Components/Admins/Form/index.js [hunk 11]
```diff
@@ -0,0 +1,158 @@
...
```
[the remark this answers is not in the document]
Comment (EmanuelMuises95) [reply src/Components/Admins/Form/index.js]: Done

File: src/Components/Admins/Form/index.js (hunk 1 above)
Comment (EmanuelMuises95) [reply src/Components/Admins/Form/index.js]: Done

File: src/Components/Admins/Modal/modal.js (hunk 3 above)
Comment (EmanuelMuises95) [reply src/Components/Admins/Modal/modal.js]: Changed

File: src/Components/Admins/Modal/modal.js (hunk 4 above)
Comment (EmanuelMuises95) [reply src/Components/Admins/Modal/modal.js]: Removed

File: src/Components/Admins/Modal/modal.js (hunk 5 above)
Comment (EmanuelMuises95) [reply src/Components/Admins/Modal/modal.js]: Removed

File: src/Components/Admins/Modal/modal.js [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+    <>
+      <div
+        id="id-screen"
+        onClick={() => setModalDisplay(false)}
+        className={modalstyles.screen}
+      ></div>
+      <div id="id-modal" className={modalstyles.modal}>
+        <header className={modalstyles.header}>
+          <h3 className={modalstyles.header__title}>{title}</h3>
+          <button className={modalstyles.header__button} onClick={() => setModalDisplay(false)}>
```
[the remark this answers is not in the document]
Comment (EmanuelMuises95) [reply src/Components/Admins/Modal/modal.js]: Done

**Review by lmalvasia — APPROVED** [review]

**Comment by Valentin-sanabria** [comment]
Hi Ema, your PR has 😥**ONE HIGH SEVERITY BUG😥** you should solve before merging. I also have things i consider important enough to change beforehand, here are all of them **in order by _most-important_ to _least-important_:**

### **QA FEEDBACK:**
🛑When deleting an Admin, the modal showing the success message **CAN NOT BE CLOSED** and the user is forced to reload the page in order to continue working. Heres a screenshot of the error, but you can reproduce this bug easily by deleting an Admin and clicking the close button in modal.

image

🟣You do not validate the existence of letters in `name` and `last name` allowing for an **EMPTY** input. I saw you validated an empty string, but, **you did not validate a string full of blank spaces**. An easy way of avoiding this type of problem would be to require a minimum amount of letters.

image

🟣 I found **_error_** messages overcomplicated. For ex here you could've said "names can't contain numbers". And in case you do not want to have a specific error message for each case, then instead of explaining what names can't have, say what they should, so in this case "name can only contain letters" would solve the issue for every input possibility.

image

A better example, is this screenshot. I genuinely **do not understand** what is wrong with this password, guessing by how other apps work I imagined it needs an uppercase but it **WAS NOT THE CASE**. So i dont know what im missing, take in mind the user should not have to guess what the error says, the error message should work as a guide for the user. Change the error messages so they're clear instructions on what to input. (later i found out it was because it didnt have 8chars)

image

🟣 When creating a new admin, and succesfully completing the inputs, the modal message says "**PROJECT** created succesfully" change it so it displays info correctly. 

image

### **💟 RECOMMENDATIONS:**
For such an early version of the app your styling is incredible. I really like how the modals look. The only thing i could point out is that:
• I think you could add a little bit of `margin-right` to this inputs and `margin-top` to the submit button and the inputs themselves so things arent as cramped together.

image

**Review by lmalvasia — APPROVED** [review]

**Pull request merged**